src/util/codec/mysql/json/comparison.rs

```diff
@@ -0,0 +1,171 @@
+                }
+            }
+
+            return match (self, right) {
+                (&Json::Boolean(left_data), &Json::Boolean(right_data)) => {
+                    left_data.partial_cmp(&right_data)
+                }
+                (&Json::String(ref left_data), &Json::String(ref right_data)) => {
+                    left_data.partial_cmp(right_data)
+                }
+                (&Json::Array(ref left_data), &Json::Array(ref right_data)) => {
+                    left_data.partial_cmp(right_data)
+                }
+                (&Json::Object(_), &Json::Object(_)) => {
+                    let mut left_data = vec![];
+                    let mut right_data = vec![];
+                    left_data.encode_json(self).unwrap();
+                    right_data.encode_json(right).unwrap();
+                    left_data.partial_cmp(&right_data)
+                }
+                _ => Some(Ordering::Equal),
+            };
+        }
+
+        let left_data = self.as_f64();
+        let right_data = right.as_f64();
+        // tidb treats boolean as integer, but boolean is different from integer in JSON.
+        // so we need convert them to same type and then compare.
+        if left_data.is_ok() && right_data.is_ok() {
+            let left = left_data.unwrap();
+            let right = right_data.unwrap();
+            return left.partial_cmp(&right);
+        }
+
+        if precedence_diff > 0 {
+            Some(Ordering::Greater)
+        } else {
+            Some(Ordering::Less)
+        }
+    }
+}
+
+#[cfg(test)]
+mod test {
+    use super::*;
+
+    #[test]
+    fn test_cmp_json_between_same_type() {
+        let test_cases = vec![
+            ("false", "true"),
+            ("3", "5"),
+            ("3.0", "4.9"),
+            (r#""hello""#, r#""hello, world""#),
+            (r#"["a", "b"]"#, r#"["a", "c"]"#),
+            (r#"{"a": "b"}"#, r#"{"a": "c"}"#),
+        ];
+        for (left_str, right_str) in test_cases {
+            let left: Json = left_str.parse().unwrap();
+            let right: Json = right_str.parse().unwrap();
+            assert!(left < right);
+            assert_eq!(left, left);
+        }
+        assert_eq!(Json::None, Json::None);
+    }
+
+    #[test]
+    fn test_cmp_json_between_diff_type() {
+        let test_cases = vec![
+            ("1.5", "2"),
+            ("1.5", "false"),
+            ("true", "1.5"),
+            ("true", "2"),
+            ("null", r#"{"a": "b"}"#),
+            ("2", r#""hello, world""#),
+            (r#""hello, world""#, r#"{"a": "b"}"#),
+            (r#"{"a": "b"}"#, r#"["a", "b"]"#),
+            (r#"["a", "b"]"#, "false"),
+        ];
+
+        for (left_str, right_str) in test_cases {
+            let left: Json = left_str.parse().unwrap();
+            let right: Json = right_str.parse().unwrap();
+            assert!(left < right);
+        }
+
+        assert_eq!(Json::I64(2), Json::Boolean(false));
```

Comment:
Here it's a natural defect of TiDB, because we treat bool as int. so if we compare bool and int, there are some incompatible cases with MySQL.